server/src/main/java/org/candlepin/dto/manifest/v1/ProductDTO.java

```diff
@@ -846,6 +847,92 @@ private boolean isNullOrIncomplete(BrandingDTO branding) {
             branding.getType() == null || branding.getType().isEmpty();
     }
 
+    /**
+     * Retrieves a view of the provided products for the product represented by this DTO.
+     * If the provided products have not yet been defined, this method returns null.
+     *
+     * Note that the collection returned by this method is a view of the collection backing this
+     * set of provided products. Elements cannot be added to the collection, but elements may be removed.
+     * Changes made to the collection will be reflected by this product DTO instance.
+     *
+     * @return
+     *  The provided products associated with this key, or null if they have not yet been defined
+     */
+    @JsonIgnore
+    @Override
+    public Set<ProductDTO> getProvidedProducts() {
+        return this.providedProducts != null ? new SetView<>(this.providedProducts) : null;
+    }
+
+    /**
+     * Adds the collection of provided products to this Product DTO.
+     *
+     * @param providedProducts
+     *  A set of provided products to attach to this DTO, or null to clear the existing ones
+     *
+     * @return
+     *  A reference to this DTO
+     */
+    @JsonIgnore
+    public ProductDTO setProvidedProducts(Set<ProductDTO> providedProducts) {
+        if (providedProducts != null) {
+            if (this.providedProducts == null) {
+                this.providedProducts = new HashSet<>();
+            }
+            else {
+                this.providedProducts.clear();
+            }
+
+            for (ProductDTO dto : providedProducts) {
```

Comment:
Would it make sense to do the validation before clearing the existing list?

Reply:
At DTO level it wouldn't make much difference as we want a fresh DTO to be returned every time with all its properties. There is still a possibility others can throw exception resulting in failure of whole operation as there are no try-catch guarding setter method invocation anywhere. Also there are multiple such methods in execution chain and failure of even one method can lead to failure of operation, which is expected.